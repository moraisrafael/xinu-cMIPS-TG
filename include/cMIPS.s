
	# see vhdl/packageMemory.vhd for addresses
        .set x_INST_BASE_ADDR,0x00000000
        .set x_INST_MEM_SZ,0x00020000

        .set x_DATA_BASE_ADDR,0x00080000
        .set x_DATA_MEM_SZ,0x00080000
	
        .set x_IO_BASE_ADDR,0x3c000000
        .set x_IO_MEM_SZ,0x00002000
	.set x_IO_ADDR_RANGE,0x00000020

	.set x_SDRAM_BASE_ADDR,0x04000000
	.set x_SDRAM_MEM_SZ,0x02000000
	
	.set HW_counter_addr,(x_IO_BASE_ADDR +  5 * x_IO_ADDR_RANGE)
	.set HW_FPU_addr,    (x_IO_BASE_ADDR +  6 * x_IO_ADDR_RANGE)
	.set HW_uart_addr,   (x_IO_BASE_ADDR +  7 * x_IO_ADDR_RANGE)
	.set HW_dsp7seg_addr,(x_IO_BASE_ADDR +  9 * x_IO_ADDR_RANGE)
	.set HW_keybd_addr,  (x_IO_BASE_ADDR + 10 * x_IO_ADDR_RANGE)
	.set HW_lcd_addr,    (x_IO_BASE_ADDR + 11 * x_IO_ADDR_RANGE)
	.set HW_sdc_addr,    (x_IO_BASE_ADDR + 12 * x_IO_ADDR_RANGE)

	# see vhdl/packageMemory.vhd for addresses
	.set x_EXCEPTION_0000,0x00000130
	.set x_EXCEPTION_0100,0x00000200
	.set x_EXCEPTION_0180,0x00000280
	.set x_EXCEPTION_0200,0x00000400
	.set x_EXCEPTION_BFC0,0x00000680
	.set x_ENTRY_POINT,   0x00000700


	.set c0_index,   $0
	.set c0_random,  $1
	.set c0_entrylo0,$2
	.set c0_entrylo1,$3
	.set c0_context ,$4
	.set c0_pagemask,$5
	.set c0_wired,   $6
	.set c0_badvaddr,$8
	.set c0_count   ,$9
	.set c0_entryhi ,$10
	.set c0_compare ,$11
	.set c0_status  ,$12
	.set c0_cause   ,$13
	.set c0_epc,     $14
	.set c0_config,  $16
	.set c0_config_f0,0
	.set c0_config_f1,1
	.set c0_lladdr,  $17
	.set c0_errorpc, $30

	
	# reset: COP0 present, at exception level, all else disabled
	.set c0_status_reset,0x10000002
	
	# normal state: COP0 present, user mode, all IRQs enabled
	.set c0_status_normal,0x1000ff11

	# reset: COUNTER stopped, use special interrVector, no interrupts
	.set c0_cause_reset, 0x0880007c


	# not used in XINU, RH
#	.equ led_RED, 0x4000
#	.equ led_GRE, 0x2000
#	.equ led_BLU, 0x1000
#	.equ led_OFF, 0x0FFF

