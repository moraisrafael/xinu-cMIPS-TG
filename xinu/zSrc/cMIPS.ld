OUTPUT_ARCH(mips)
ENTRY(_start)

MEMORY
{
        rom (rx)  : ORIGIN = 0x, /* x_INST_BASE_ADDR */
                    LENGTH = 0x, /* x_INST_MEM_SZ */
        ram (!rx) : ORIGIN = 0x, /* x_DATA_BASE_ADDR */
                    LENGTH = 0x, /* x_DATA_MEM_SZ */
        io (!rx)  : ORIGIN = 0x0f000000, /* not used, do not remove       */
                    LENGTH = 0x00020000  /* edMemory.sh needs thess lines */
}

SECTIONS
{
       .text :
       {
          *(.text .text.*)
          _etext = . ;         /* end of text constant (from Xinu) */
       }                                   > rom

       .data : 
       {
          _bdata = . ;         /* begin of data constant */
          *(.data .data.*) 
          _edata = . ;         /* end of data constant  (from Xinu) */
       }                                   > ram
       .data1   : { *(.data1) }            > ram
       .rodata  : { *(.rodata .rodata.*) } > ram
       .rodata1 : { *(.rodata1) }          > ram
       .lit8    : { *(.lit8) }             > ram
       .lit4    : { *(.lit4) }             > ram
       .sdata   : { *(.sdata .sdata.*) }   > ram
       .sbss    : { *(.sbss .sbss.*) *(.scommon .scommon.*) }  > ram
       .bss :
       {
          *(.bss .bss.*) *(COMMON) 
          _end = . ;           /* end of image constant  (from Xinu) */
       }                                   > ram

       end_RAM = 0x; /* x_DATA_MEM_SZ */
       half_RAM = (end_RAM / 2);
       base_TP = ( _bdata + half_RAM );
       .TP base_TP : { *(.TP) }                  > ram
}
