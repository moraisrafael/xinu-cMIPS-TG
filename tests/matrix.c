/*
/* matrix_m.c
/*
/* by:
/* Brian Grattan
/*
/**********************************************************************/
#ifdef cMIPS
  #include "cMIPS.h"
#else
  #include <stdio.h>
#endif
/**********************************************************************/


int random(int *m_z, int *m_w) {   // from wikipedia
  *m_z = 36969 * (*m_z & 65535) + (*m_z >> 16);
  *m_w = 18000 * (*m_w & 65535) + (*m_w >> 16);
  return ((*m_z << 16) + *m_w);  /* 32-bit result */
}

#define mSz 4

void main () {

  int matrix_a [mSz][mSz]; // = { { 1, 2, 3 }, { 4, 5,-6 }, {-7, 8, 9 } };
  int matrix_b [mSz][mSz]; // = { {-1, 1,-6 }, {-2, 2, 8 }, {-3, 3, 4 } };
  int result [mSz][mSz];
  int i, j, k;
  int sum;

  int m_w = 17;    /* must not be zero, nor 0x464fffff */
  int m_z = 31;    /* must not be zero, nor 0x9068ffff */


  for ( i=0; i<mSz; i++ ) {
    for ( j=0; j<mSz; j++ ) {
      matrix_a[i][j] = random(&m_z, &m_w);
      matrix_b[i][j] = random(&m_z, &m_w);
    }
  }

#ifdef cMIPS
  int *IO = (int *)x_IO_BASE_ADDR;
#endif
	
  for ( i=0; i<mSz; i++ ) {
    for ( j=0; j<mSz; j++ ) {
      sum = 0;
      for ( k=0; k<mSz; k++ ) {
	sum = sum + ( matrix_a [i][k] * matrix_b [k][j] );
	// *IO = sum;
      }
      result [i][j] = sum;
    } /* end j */
  } /* end i */

  for ( i=0; i<mSz; i++ ) {
    for ( j=0; j<mSz; j++ ) {

#ifdef cMIPS
      print(result[i][j]);
#else
      printf("%08x\n",result[i][j]);
#endif

    } /* end j */

#ifdef cMIPS
    to_stdout('\n');
#else
    printf("\n");
#endif
  } /* end i */

} /* end main */

