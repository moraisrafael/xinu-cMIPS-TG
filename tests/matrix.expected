b58873f2
f43f6140
af2271b0
f5911983

aa89ab82
294741b6
c406f496
0089c07e

be10044a
f65b2e96
ff2932d6
2bed782c

9618ad98
37cf358a
fa7ae06a
edb03e79

